File: BZEAD-iOS-main/BZEAD/Data/ProfileRepository.swift
import Foundation

final class ProfileRepository {
    private let http = SupabaseHTTP.shared

    func fetchProfile(session: BuyerSession) async -> ProfileDetail? {
        guard SupabaseConfig.isConfigured() else { return nil }

        let url = "\(SupabaseConfig.url)/rest/v1/profiles" +
            "?id=eq.\(session.userId)" +
            "&select=id,full_name,email,phone,country_id,notification_preferences"

        do {
            let (data, response) = try await http.getAllowingErrorStatus(url, headers: http.authHeaders(session: session))
            guard http.isSuccess(response) else { return nil }
            return try http.decode([ProfileDetail].self, from: data).first
        } catch {
            return nil
        }
    }

    func updateProfile(
        session: BuyerSession,
        fullName: String?,
        phone: String?,
        countryId: String? = nil
    ) async -> Result<Void, Error> {
        var body: [String: Any] = [:]
        if let fullName { body["full_name"] = fullName }
        if let phone { body["phone"] = phone }
        if let countryId { body["country_id"] = countryId }

        var headers = http.authHeaders(session: session)
        headers["Content-Type"] = "application/json"

        do {
            let (data, response) = try await http.patchAllowingErrorStatus(
                "\(SupabaseConfig.url)/rest/v1/profiles?id=eq.\(session.userId)",
                headers: headers,
                body: try http.encodeJSONObject(body)
            )
            return http.isSuccess(response)
                ? .success(())
                : .failure(NSError(domain: "ProfileRepository", code: 1, userInfo: [
                    NSLocalizedDescriptionKey: http.bodyAsText(data).isEmpty ? "Failed to update profile" : http.bodyAsText(data),
                ]))
        } catch {
            return .failure(error)
        }
    }

    func updateNotificationPreferences(
        session: BuyerSession,
        preferences: [String: JSONValue]
    ) async -> Result<Void, Error> {
        let body = try? http.encoder.encode(["notification_preferences": preferences])
        var headers = http.authHeaders(session: session)
        headers["Content-Type"] = "application/json"

        do {
            let (data, response) = try await http.patchAllowingErrorStatus(
                "\(SupabaseConfig.url)/rest/v1/profiles?id=eq.\(session.userId)",
                headers: headers,
                body: body ?? Data()
            )
            return http.isSuccess(response)
                ? .success(())
                : .failure(NSError(domain: "ProfileRepository", code: 2, userInfo: [
                    NSLocalizedDescriptionKey: http.bodyAsText(data).isEmpty ? "Failed to update notifications" : http.bodyAsText(data),
                ]))
        } catch {
            return .failure(error)
        }
    }

    func changePassword(
        session: BuyerSession,
        currentPassword: String,
        newPassword: String
    ) async -> Result<Void, Error> {
        var verifyHeaders = http.anonHeaders()
        verifyHeaders["Content-Type"] = "application/json"
        let verifyBody = try? http.encodeJSONObject([
            "email": session.email,
            "password": currentPassword,
        ])

        do {
            let (verifyData, verifyResponse) = try await http.postAllowingErrorStatus(
                "\(SupabaseConfig.url)/auth/v1/token?grant_type=password",
                headers: verifyHeaders,
                body: verifyBody
            )
            guard http.isSuccess(verifyResponse) else {
                return .failure(NSError(domain: "ProfileRepository", code: 3, userInfo: [
                    NSLocalizedDescriptionKey: "Current password is incorrect",
                ]))
            }
            _ = verifyData

            var headers = http.anonHeaders()
            headers["Authorization"] = "Bearer \(session.accessToken)"
            headers["Content-Type"] = "application/json"
            let (data, response) = try await http.putAllowingErrorStatus(
                "\(SupabaseConfig.url)/auth/v1/user",
                headers: headers,
                body: try http.encodeJSONObject(["password": newPassword])
            )
            return http.isSuccess(response)
                ? .success(())
                : .failure(NSError(domain: "ProfileRepository", code: 4, userInfo: [
                    NSLocalizedDescriptionKey: http.bodyAsText(data).isEmpty ? "Failed to change password" : http.bodyAsText(data),
                ]))
        } catch {
            return .failure(error)
        }
    }

    func changePasswordWithoutVerify(session: BuyerSession, newPassword: String) async -> Result<Void, Error> {
        var headers = http.anonHeaders()
        headers["Authorization"] = "Bearer \(session.accessToken)"
        headers["Content-Type"] = "application/json"
        do {
            let (data, response) = try await http.putAllowingErrorStatus(
                "\(SupabaseConfig.url)/auth/v1/user",
                headers: headers,
                body: try http.encodeJSONObject(["password": newPassword])
            )
            return http.isSuccess(response)
                ? .success(())
                : .failure(NSError(domain: "ProfileRepository", code: 5, userInfo: [
                    NSLocalizedDescriptionKey: http.bodyAsText(data).isEmpty ? "Failed to change password" : http.bodyAsText(data),
                ]))
        } catch {
            return .failure(error)
        }
    }
}

final class AddressRepository {
    private let http = SupabaseHTTP.shared

    func fetchAddresses(session: BuyerSession) async -> [UserAddressRow] {
        guard SupabaseConfig.isConfigured() else { return [] }

        let url = "\(SupabaseConfig.url)/rest/v1/user_addresses" +
            "?user_id=eq.\(session.userId)" +
            "&select=*" +
            "&order=is_default.desc"

        do {
            let (data, response) = try await http.getAllowingErrorStatus(url, headers: http.authHeaders(session: session))
            guard http.isSuccess(response) else { return [] }
            return try http.decode([UserAddressRow].self, from: data)
        } catch {
            return []
        }
    }

    func createAddress(session: BuyerSession, address: UserAddressRow) async -> Result<UserAddressRow, Error> {
        let body: [String: Any] = [
            "user_id": session.userId,
            "full_name": address.fullName,
            "phone_number": address.phoneNumber as Any,
            "email": address.email as Any,
            "country": address.country,
            "street_address_1": address.streetAddress1,
            "street_address_2": address.streetAddress2 as Any,
            "city": address.city,
            "state": address.state,
            "postal_code": address.postalCode,
            "address_type": address.addressType,
            "delivery_notes": address.deliveryNotes as Any,
            "is_default": address.isDefault,
        ]

        var headers = http.authHeaders(session: session)
        headers["Prefer"] = "return=representation"
        headers["Content-Type"] = "application/json"

        do {
            let (data, response) = try await http.postJSONObjectAllowingErrorStatus(
                "\(SupabaseConfig.url)/rest/v1/user_addresses",
                headers: headers,
                body: body
            )
            guard http.isSuccess(response) else {
                return .failure(NSError(domain: "AddressRepository", code: 1, userInfo: [
                    NSLocalizedDescriptionKey: http.bodyAsText(data).isEmpty ? "Failed to create address" : http.bodyAsText(data),
                ]))
            }
            guard let created = try http.decode([UserAddressRow].self, from: data).first else {
                return .failure(NSError(domain: "AddressRepository", code: 2, userInfo: [
                    NSLocalizedDescriptionKey: "No address returned",
                ]))
            }
            if address.isDefault {
                _ = await setDefault(session: session, addressId: created.id)
            }
            return .success(created)
        } catch {
            return .failure(error)
        }
    }

    func updateAddress(session: BuyerSession, addressId: String, address: UserAddressRow) async -> Result<Void, Error> {
        var headers = http.authHeaders(session: session)
        headers["Content-Type"] = "application/json"
        do {
            let body = try http.encodeBody(address)
            let (data, response) = try await http.patchAllowingErrorStatus(
                "\(SupabaseConfig.url)/rest/v1/user_addresses?id=eq.\(addressId)&user_id=eq.\(session.userId)",
                headers: headers,
                body: body
            )
            return http.isSuccess(response)
                ? .success(())
                : .failure(NSError(domain: "AddressRepository", code: 3, userInfo: [
                    NSLocalizedDescriptionKey: http.bodyAsText(data).isEmpty ? "Failed to update address" : http.bodyAsText(data),
                ]))
        } catch {
            return .failure(error)
        }
    }

    func deleteAddress(session: BuyerSession, addressId: String) async -> Result<Void, Error> {
        do {
            let (data, response) = try await http.deleteAllowingErrorStatus(
                "\(SupabaseConfig.url)/rest/v1/user_addresses?id=eq.\(addressId)&user_id=eq.\(session.userId)",
                headers: http.authHeaders(session: session)
            )
            return http.isSuccess(response)
                ? .success(())
                : .failure(NSError(domain: "AddressRepository", code: 4, userInfo: [
                    NSLocalizedDescriptionKey: http.bodyAsText(data).isEmpty ? "Failed to delete address" : http.bodyAsText(data),
                ]))
        } catch {
            return .failure(error)
        }
    }

    func setDefault(session: BuyerSession, addressId: String) async -> Result<Void, Error> {
        var headers = http.authHeaders(session: session)
        headers["Content-Type"] = "application/json"
        do {
            let (patchData, patchResponse) = try await http.patchAllowingErrorStatus(
                "\(SupabaseConfig.url)/rest/v1/user_addresses?id=eq.\(addressId)&user_id=eq.\(session.userId)",
                headers: headers,
                body: try http.encodeJSONObject(["is_default": true])
            )
            guard http.isSuccess(patchResponse) else {
                return .failure(NSError(domain: "AddressRepository", code: 5, userInfo: [
                    NSLocalizedDescriptionKey: http.bodyAsText(patchData).isEmpty ? "Failed to set default address" : http.bodyAsText(patchData),
                ]))
            }

            _ = try? await http.patchAllowingErrorStatus(
                "\(SupabaseConfig.url)/rest/v1/user_addresses?user_id=eq.\(session.userId)&id=neq.\(addressId)&is_default=eq.true",
                headers: headers,
                body: try http.encodeJSONObject(["is_default": false])
            )
            return .success(())
        } catch {
            return .failure(error)
        }
    }
}

File: BZEAD-APK-main/app/src/main/java/com/bzead/apk/data/ProfileRepository.kt
package com.bzead.apk.data

import com.bzead.apk.data.model.BuyerSession
import com.bzead.apk.data.model.ProfileDetail
import com.bzead.apk.data.model.UserAddressRow
import io.ktor.client.call.body
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.patch
import io.ktor.client.request.post
import io.ktor.client.request.put
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import com.bzead.apk.data.SupabaseHttp.authHeaders
import io.ktor.http.isSuccess
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

class ProfileRepository {
    suspend fun fetchProfile(session: BuyerSession): ProfileDetail? {
        if (!SupabaseConfig.isConfigured()) return null

        val response = SupabaseHttp.client.get(
            "${SupabaseConfig.url}/rest/v1/profiles" +
                "?id=eq.${session.userId}" +
                "&select=id,full_name,email,phone,country_id,notification_preferences",
        ) {
            authHeaders(session)
        }
        if (!response.status.isSuccess()) return null
        return response.body<List<ProfileDetail>>().firstOrNull()
    }

    suspend fun updateProfile(
        session: BuyerSession,
        fullName: String?,
        phone: String?,
        countryId: String? = null,
        currency: String? = null,
    ): Result<Unit> {
        val body = buildJsonObject {
            fullName?.let { put("full_name", it) }
            phone?.let { put("phone", it) }
            countryId?.let { put("country_id", it) }
            currency?.let { put("currency", it) }
        }
        val response = SupabaseHttp.client.patch(
            "${SupabaseConfig.url}/rest/v1/profiles?id=eq.${session.userId}",
        ) {
            authHeaders(session)
            contentType(ContentType.Application.Json)
            setBody(body)
        }
        return if (response.status.isSuccess()) Result.success(Unit)
        else Result.failure(IllegalStateException("Failed to update profile"))
    }

    suspend fun updateNotificationPreferences(
        session: BuyerSession,
        preferences: JsonObject,
    ): Result<Unit> {
        val body = buildJsonObject { put("notification_preferences", preferences) }
        val response = SupabaseHttp.client.patch(
            "${SupabaseConfig.url}/rest/v1/profiles?id=eq.${session.userId}",
        ) {
            authHeaders(session)
            contentType(ContentType.Application.Json)
            setBody(body)
        }
        return if (response.status.isSuccess()) Result.success(Unit)
        else Result.failure(IllegalStateException("Failed to update notifications"))
    }

    suspend fun changePassword(
        session: BuyerSession,
        currentPassword: String,
        newPassword: String,
    ): Result<Unit> {
        val verify = SupabaseHttp.client.post("${SupabaseConfig.url}/auth/v1/token?grant_type=password") {
            header("apikey", SupabaseConfig.anonKey)
            contentType(ContentType.Application.Json)
            setBody(
                buildJsonObject {
                    put("email", session.email)
                    put("password", currentPassword)
                },
            )
        }
        if (!verify.status.isSuccess()) {
            return Result.failure(IllegalStateException("Current password is incorrect"))
        }
        val response = SupabaseHttp.client.put("${SupabaseConfig.url}/auth/v1/user") {
            header("apikey", SupabaseConfig.anonKey)
            header(HttpHeaders.Authorization, "Bearer ${session.accessToken}")
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("password", newPassword) })
        }
        return if (response.status.isSuccess()) Result.success(Unit)
        else Result.failure(IllegalStateException("Failed to change password"))
    }

    suspend fun changePasswordWithoutVerify(
        session: BuyerSession,
        newPassword: String,
    ): Result<Unit> {
        val response = SupabaseHttp.client.put("${SupabaseConfig.url}/auth/v1/user") {
            header("apikey", SupabaseConfig.anonKey)
            header(HttpHeaders.Authorization, "Bearer ${session.accessToken}")
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("password", newPassword) })
        }
        return if (response.status.isSuccess()) Result.success(Unit)
        else Result.failure(IllegalStateException("Failed to change password"))
    }
}

class AddressRepository {
    suspend fun fetchAddresses(session: BuyerSession): List<UserAddressRow> {
        if (!SupabaseConfig.isConfigured()) return emptyList()

        val response = SupabaseHttp.client.get(
            "${SupabaseConfig.url}/rest/v1/user_addresses" +
                "?user_id=eq.${session.userId}" +
                "&select=*" +
                "&order=is_default.desc",
        ) {
            authHeaders(session)
        }
        return if (response.status.isSuccess()) response.body() else emptyList()
    }

    suspend fun createAddress(session: BuyerSession, address: UserAddressRow): Result<UserAddressRow> {
        val body = buildJsonObject {
            put("user_id", session.userId)
            put("full_name", address.fullName)
            put("phone_number", address.phoneNumber)
            put("email", address.email)
            put("country", address.country)
            put("street_address_1", address.streetAddress1)
            put("street_address_2", address.streetAddress2)
            put("city", address.city)
            put("state", address.state)
            put("postal_code", address.postalCode)
            put("address_type", address.addressType)
            put("delivery_notes", address.deliveryNotes)
            put("is_default", address.isDefault)
        }
        val response = SupabaseHttp.client.post("${SupabaseConfig.url}/rest/v1/user_addresses") {
            authHeaders(session)
            header("Prefer", "return=representation")
            contentType(ContentType.Application.Json)
            setBody(body)
        }
        if (!response.status.isSuccess()) {
            return Result.failure(IllegalStateException("Failed to create address"))
        }
        val created = response.body<List<UserAddressRow>>().firstOrNull()
            ?: return Result.failure(IllegalStateException("No address returned"))
        if (address.isDefault) setDefault(session, created.id)
        return Result.success(created)
    }

    suspend fun updateAddress(
        session: BuyerSession,
        addressId: String,
        address: UserAddressRow,
    ): Result<Unit> {
        val response = SupabaseHttp.client.patch(
            "${SupabaseConfig.url}/rest/v1/user_addresses?id=eq.$addressId&user_id=eq.${session.userId}",
        ) {
            authHeaders(session)
            contentType(ContentType.Application.Json)
            setBody(address)
        }
        return if (response.status.isSuccess()) Result.success(Unit)
        else Result.failure(IllegalStateException("Failed to update address"))
    }

    suspend fun deleteAddress(session: BuyerSession, addressId: String): Result<Unit> {
        val response = SupabaseHttp.client.delete(
            "${SupabaseConfig.url}/rest/v1/user_addresses?id=eq.$addressId&user_id=eq.${session.userId}",
        ) {
            authHeaders(session)
        }
        return if (response.status.isSuccess()) Result.success(Unit)
        else Result.failure(IllegalStateException("Failed to delete address"))
    }

    suspend fun setDefault(session: BuyerSession, addressId: String): Result<Unit> {
        val patch = SupabaseHttp.client.patch(
            "${SupabaseConfig.url}/rest/v1/user_addresses?id=eq.$addressId&user_id=eq.${session.userId}",
        ) {
            authHeaders(session)
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("is_default", true) })
        }
        if (!patch.status.isSuccess()) {
            return Result.failure(IllegalStateException("Failed to set default address"))
        }
        SupabaseHttp.client.patch(
            "${SupabaseConfig.url}/rest/v1/user_addresses?user_id=eq.${session.userId}&id=neq.$addressId&is_default=eq.true",
        ) {
            authHeaders(session)
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("is_default", false) })
        }
        return Result.success(Unit)
    }
}
